### ISSUE TITLE
0.9.0 fails on Ruby 2.5

### ISSUE THREAD

**Issue by vassilevsky** [rubysec/bundler-audit#319]
Hi :)

Not sure if Ruby 2.5 is in your supported version range.

If it's not, feel free to close this. I know I should upgrade.

If it is, then here's the error:

```
$ gem install bundler-audit bundler-leak brakeman
Successfully installed thor-1.1.0
Successfully installed bundler-audit-0.9.0
Parsing documentation for thor-1.1.0
Installing ri documentation for thor-1.1.0
Parsing documentation for bundler-audit-0.9.0
Installing ri documentation for bundler-audit-0.9.0
Done installing documentation for thor, bundler-audit after 2 seconds
Successfully installed bundler-leak-0.2.0
Parsing documentation for bundler-leak-0.2.0
Installing ri documentation for bundler-leak-0.2.0
Done installing documentation for bundler-leak after 0 seconds
Successfully installed brakeman-5.1.1
Parsing documentation for brakeman-5.1.1
Installing ri documentation for brakeman-5.1.1
Done installing documentation for brakeman after 3 seconds
4 gems installed
$ rbenv rehash
$ bundler-audit check --update
Download ruby-advisory-db ...
Cloning into '/root/.local/share/ruby-advisory-db'...
ruby-advisory-db:
  advisories:	516 advisories
  last updated:	2021-08-31 01:50:40 -0700
/usr/local/rbenv/versions/2.5.3-p0/lib/ruby/2.5.0/psych.rb:313:in `safe_load': unknown keyword: permitted_classes (ArgumentError)
	from /usr/local/rbenv/versions/2.5.3-p0/lib/ruby/gems/2.5.0/gems/bundler-audit-0.9.0/lib/bundler/audit/advisory.rb:52:in `block in load'
	from /usr/local/rbenv/versions/2.5.3-p0/lib/ruby/gems/2.5.0/gems/bundler-audit-0.9.0/lib/bundler/audit/advisory.rb:51:in `open'
	from /usr/local/rbenv/versions/2.5.3-p0/lib/ruby/gems/2.5.0/gems/bundler-audit-0.9.0/lib/bundler/audit/advisory.rb:51:in `load'
	from /usr/local/rbenv/versions/2.5.3-p0/lib/ruby/gems/2.5.0/gems/bundler-audit-0.9.0/lib/bundler/audit/database.rb:287:in `block in advisories_for'
	from /usr/local/rbenv/versions/2.5.3-p0/lib/ruby/gems/2.5.0/gems/bundler-audit-0.9.0/lib/bundler/audit/database.rb:375:in `glob'
	from /usr/local/rbenv/versions/2.5.3-p0/lib/ruby/gems/2.5.0/gems/bundler-audit-0.9.0/lib/bundler/audit/database.rb:375:in `each_advisory_path_for'
	from /usr/local/rbenv/versions/2.5.3-p0/lib/ruby/gems/2.5.0/gems/bundler-audit-0.9.0/lib/bundler/audit/database.rb:286:in `advisories_for'
	from /usr/local/rbenv/versions/2.5.3-p0/lib/ruby/gems/2.5.0/gems/bundler-audit-0.9.0/lib/bundler/audit/database.rb:310:in `check_gem'
	from /usr/local/rbenv/versions/2.5.3-p0/lib/ruby/gems/2.5.0/gems/bundler-audit-0.9.0/lib/bundler/audit/scanner.rb:225:in `block in scan_specs'
	from /usr/local/rbenv/versions/2.5.3-p0/lib/ruby/gems/2.5.0/gems/bundler-audit-0.9.0/lib/bundler/audit/scanner.rb:224:in `each'
	from /usr/local/rbenv/versions/2.5.3-p0/lib/ruby/gems/2.5.0/gems/bundler-audit-0.9.0/lib/bundler/audit/scanner.rb:224:in `scan_specs'
	from /usr/local/rbenv/versions/2.5.3-p0/lib/ruby/gems/2.5.0/gems/bundler-audit-0.9.0/lib/bundler/audit/scanner.rb:149:in `scan'
	from /usr/local/rbenv/versions/2.5.3-p0/lib/ruby/gems/2.5.0/gems/bundler-audit-0.9.0/lib/bundler/audit/scanner.rb:119:in `report'
	from /usr/local/rbenv/versions/2.5.3-p0/lib/ruby/gems/2.5.0/gems/bundler-audit-0.9.0/lib/bundler/audit/cli.rb:76:in `check'
	from /usr/local/rbenv/versions/2.5.3-p0/lib/ruby/gems/2.5.0/gems/thor-1.1.0/lib/thor/command.rb:27:in `run'
	from /usr/local/rbenv/versions/2.5.3-p0/lib/ruby/gems/2.5.0/gems/thor-1.1.0/lib/thor/invocation.rb:127:in `invoke_command'
	from /usr/local/rbenv/versions/2.5.3-p0/lib/ruby/gems/2.5.0/gems/thor-1.1.0/lib/thor.rb:392:in `dispatch'
	from /usr/local/rbenv/versions/2.5.3-p0/lib/ruby/gems/2.5.0/gems/thor-1.1.0/lib/thor/base.rb:485:in `start'
	from /usr/local/rbenv/versions/2.5.3-p0/lib/ruby/gems/2.5.0/gems/bundler-audit-0.9.0/bin/bundle-audit:10:in `<top (required)>'
	from /usr/local/rbenv/versions/2.5.3-p0/lib/ruby/gems/2.5.0/gems/bundler-audit-0.9.0/bin/bundler-audit:3:in `load'
	from /usr/local/rbenv/versions/2.5.3-p0/lib/ruby/gems/2.5.0/gems/bundler-audit-0.9.0/bin/bundler-audit:3:in `<top (required)>'
	from /usr/local/rbenv/versions/2.5.3/bin/bundler-audit:23:in `load'
	from /usr/local/rbenv/versions/2.5.3/bin/bundler-audit:23:in `<main>'
```

It would be cool to be able to run on 2.5 :)

**Comment by ericpetroelje** [comment]
Having a similar, but slightly different error with Ruby 2.3:
```


gem install bundler-audit -v '>= 0.8.0.rc1'
bundle-audit check --update
Successfully installed thor-1.1.0
Successfully installed bundler-audit-0.9.0
2 gems installed
Download ruby-advisory-db ...
Cloning into '/home/runner/.local/share/ruby-advisory-db'...
ruby-advisory-db:
  advisories:	516 advisories
  last updated:	2021-08-31 01:50:40 -0700
/opt/hostedtoolcache/Ruby/2.3.6/x64/lib/ruby/2.3.0/psych/class_loader.rb:97:in `find': Tried to load unspecified class: Date (Psych::DisallowedClass)
	from /opt/hostedtoolcache/Ruby/2.3.6/x64/lib/ruby/2.3.0/psych/class_loader.rb:28:in `load'
	from /opt/hostedtoolcache/Ruby/2.3.6/x64/lib/ruby/2.3.0/psych/class_loader.rb:39:in `block (2 levels) in <class:ClassLoader>'
	from /opt/hostedtoolcache/Ruby/2.3.6/x64/lib/ruby/2.3.0/psych/scalar_scanner.rb:70:in `tokenize'
	from /opt/hostedtoolcache/Ruby/2.3.6/x64/lib/ruby/2.3.0/psych/visitors/to_ruby.rb:60:in `deserialize'
	from /opt/hostedtoolcache/Ruby/2.3.6/x64/lib/ruby/2.3.0/psych/visitors/to_ruby.rb:123:in `visit_Psych_Nodes_Scalar'
	from /opt/hostedtoolcache/Ruby/2.3.6/x64/lib/ruby/2.3.0/psych/visitors/visitor.rb:16:in `visit'
	from /opt/hostedtoolcache/Ruby/2.3.6/x64/lib/ruby/2.3.0/psych/visitors/visitor.rb:6:in `accept'
	from /opt/hostedtoolcache/Ruby/2.3.6/x64/lib/ruby/2.3.0/psych/visitors/to_ruby.rb:32:in `accept'
	from /opt/hostedtoolcache/Ruby/2.3.6/x64/lib/ruby/2.3.0/psych/visitors/to_ruby.rb:338:in `block in revive_hash'
	from /opt/hostedtoolcache/Ruby/2.3.6/x64/lib/ruby/2.3.0/psych/visitors/to_ruby.rb:336:in `each'
	from /opt/hostedtoolcache/Ruby/2.3.6/x64/lib/ruby/2.3.0/psych/visitors/to_ruby.rb:336:in `each_slice'
	from /opt/hostedtoolcache/Ruby/2.3.6/x64/lib/ruby/2.3.0/psych/visitors/to_ruby.rb:336:in `revive_hash'
	from /opt/hostedtoolcache/Ruby/2.3.6/x64/lib/ruby/2.3.0/psych/visitors/to_ruby.rb:162:in `visit_Psych_Nodes_Mapping'
	from /opt/hostedtoolcache/Ruby/2.3.6/x64/lib/ruby/2.3.0/psych/visitors/visitor.rb:16:in `visit'
	from /opt/hostedtoolcache/Ruby/2.3.6/x64/lib/ruby/2.3.0/psych/visitors/visitor.rb:6:in `accept'
	from /opt/hostedtoolcache/Ruby/2.3.6/x64/lib/ruby/2.3.0/psych/visitors/to_ruby.rb:32:in `accept'
	from /opt/hostedtoolcache/Ruby/2.3.6/x64/lib/ruby/2.3.0/psych/visitors/to_ruby.rb:311:in `visit_Psych_Nodes_Document'
	from /opt/hostedtoolcache/Ruby/2.3.6/x64/lib/ruby/2.3.0/psych/visitors/visitor.rb:16:in `visit'
	from /opt/hostedtoolcache/Ruby/2.3.6/x64/lib/ruby/2.3.0/psych/visitors/visitor.rb:6:in `accept'
	from /opt/hostedtoolcache/Ruby/2.3.6/x64/lib/ruby/2.3.0/psych/visitors/to_ruby.rb:32:in `accept'
	from /opt/hostedtoolcache/Ruby/2.3.6/x64/lib/ruby/2.3.0/psych.rb:304:in `safe_load'
	from /opt/hostedtoolcache/Ruby/2.3.6/x64/lib/ruby/gems/2.3.0/gems/bundler-audit-0.9.0/lib/bundler/audit/advisory.rb:52:in `block in load'
	from /opt/hostedtoolcache/Ruby/2.3.6/x64/lib/ruby/gems/2.3.0/gems/bundler-audit-0.9.0/lib/bundler/audit/advisory.rb:51:in `open'
	from /opt/hostedtoolcache/Ruby/2.3.6/x64/lib/ruby/gems/2.3.0/gems/bundler-audit-0.9.0/lib/bundler/audit/advisory.rb:51:in `load'
	from /opt/hostedtoolcache/Ruby/2.3.6/x64/lib/ruby/gems/2.3.0/gems/bundler-audit-0.9.0/lib/bundler/audit/database.rb:287:in `block in advisories_for'
	from /opt/hostedtoolcache/Ruby/2.3.6/x64/lib/ruby/gems/2.3.0/gems/bundler-audit-0.9.0/lib/bundler/audit/database.rb:375:in `glob'
	from /opt/hostedtoolcache/Ruby/2.3.6/x64/lib/ruby/gems/2.3.0/gems/bundler-audit-0.9.0/lib/bundler/audit/database.rb:375:in `each_advisory_path_for'
	from /opt/hostedtoolcache/Ruby/2.3.6/x64/lib/ruby/gems/2.3.0/gems/bundler-audit-0.9.0/lib/bundler/audit/database.rb:286:in `advisories_for'
	from /opt/hostedtoolcache/Ruby/2.3.6/x64/lib/ruby/gems/2.3.0/gems/bundler-audit-0.9.0/lib/bundler/audit/database.rb:310:in `check_gem'
	from /opt/hostedtoolcache/Ruby/2.3.6/x64/lib/ruby/gems/2.3.0/gems/bundler-audit-0.9.0/lib/bundler/audit/scanner.rb:225:in `block in scan_specs'
	from /opt/hostedtoolcache/Ruby/2.3.6/x64/lib/ruby/gems/2.3.0/gems/bundler-audit-0.9.0/lib/bundler/audit/scanner.rb:224:in `each'
	from /opt/hostedtoolcache/Ruby/2.3.6/x64/lib/ruby/gems/2.3.0/gems/bundler-audit-0.9.0/lib/bundler/audit/scanner.rb:224:in `scan_specs'
	from /opt/hostedtoolcache/Ruby/2.3.6/x64/lib/ruby/gems/2.3.0/gems/bundler-audit-0.9.0/lib/bundler/audit/scanner.rb:149:in `scan'
	from /opt/hostedtoolcache/Ruby/2.3.6/x64/lib/ruby/gems/2.3.0/gems/bundler-audit-0.9.0/lib/bundler/audit/scanner.rb:119:in `report'
	from /opt/hostedtoolcache/Ruby/2.3.6/x64/lib/ruby/gems/2.3.0/gems/bundler-audit-0.9.0/lib/bundler/audit/cli.rb:76:in `check'
	from /opt/hostedtoolcache/Ruby/2.3.6/x64/lib/ruby/gems/2.3.0/gems/thor-1.1.0/lib/thor/command.rb:27:in `run'
	from /opt/hostedtoolcache/Ruby/2.3.6/x64/lib/ruby/gems/2.3.0/gems/thor-1.1.0/lib/thor/invocation.rb:127:in `invoke_command'
	from /opt/hostedtoolcache/Ruby/2.3.6/x64/lib/ruby/gems/2.3.0/gems/thor-1.1.0/lib/thor.rb:392:in `dispatch'
	from /opt/hostedtoolcache/Ruby/2.3.6/x64/lib/ruby/gems/2.3.0/gems/thor-1.1.0/lib/thor/base.rb:485:in `start'
	from /opt/hostedtoolcache/Ruby/2.3.6/x64/lib/ruby/gems/2.3.0/gems/bundler-audit-0.9.0/bin/bundle-audit:10:in `<top (required)>'
	from /opt/hostedtoolcache/Ruby/2.3.6/x64/bin/bundle-audit:22:in `load'
	from /opt/hostedtoolcache/Ruby/2.3.6/x64/bin/bundle-audit:22:in `<main>'
```

**Comment by postmodern** [comment]
I wasn't aware so many people were still using older Ruby versions. Note that Ruby 2.5 was [EoLed back in March of this year and Ruby 2.3 EoLed back in 2019!](https://www.ruby-lang.org/en/downloads/branches/). Ruby 2.5 and 2.3 are no longer receiving any security patches.

I pushed an ugly workaround to detect Psych < 3.1.0 and use the old calling convention (`YAML.safe_load(data, [Data])`). 8a3a1972f916d1487004014410d1b9c52c9e3f2a

**Comment by postmodern** [comment]
GitHub Actions confirmed the fix works on Ruby 2.5. https://github.com/rubysec/bundler-audit/runs/3476950230?check_suite_focus=true

**Comment by vassilevsky** [comment]
Many thanks! Yes, I know I should upgrade Ruby :)

**Comment by postmodern** [comment]
bundler-audit 0.9.0.1 released! I was able to test against ruby-2.5.9 and confirmed it works, but I wasn't even able to compile ruby-2.3.8 due to openssl compiler errors. If you're still using ruby-2.3.x, seriously consider upgrading to 2.5.x or 2.6.x.
